editor/test/integration/save_test.clj

```diff
@@ -119,21 +93,32 @@
     (with-clean-system
       (let [workspace (test-util/setup-workspace! world)
             project (test-util/setup-project! workspace)
-            save-data (group-by :resource (project/all-save-data project))]
+            save-data-by-resource (into {}
+                                        (map (juxt :resource identity))
+                                        (project/all-save-data project))]
         (doseq [query queries]
           (testing (format "Saving %s" query)
-            (let [[resource _] (first (project/find-resources project query))
-                  save (first (get save-data resource))
-                  file (slurp resource)
-                  pb-class (-> resource resource/resource-type :ext ext->proto)]
-              (if pb-class
-                (let [pb-save (protobuf/read-text pb-class (StringReader. (:content save)))
-                      pb-disk (protobuf/read-text pb-class resource)
-                      path []
-                      [disk save both] (data/diff (get-in pb-disk path) (get-in pb-save path))]
-                  (is (nil? disk))
-                  (is (nil? save)))
-                (is (= file (:content save)))))))))))
+            (let [resource (ffirst (project/find-resources project query))
+                  resource-type (resource/resource-type resource)
+                  save-string (:content (save-data-by-resource resource))]
+              (if-some [read-fn (:read-fn resource-type)]
+
+                ;; We have a read-fn. Compare the read data representations.
+                (let [disk-pb-msg (read-fn resource)
+                      save-pb-msg (with-open [reader (StringReader. save-string)]
+                                    (read-fn reader))
+                      [only-in-disk-pb-msg only-in-save-pb-msg] (data/diff disk-pb-msg save-pb-msg)]
+                  (is (nil? only-in-disk-pb-msg))
+                  (is (nil? only-in-save-pb-msg))
+                  (when (or (some? only-in-disk-pb-msg)
+                            (some? only-in-save-pb-msg))
+                    (println "When comparing" (resource/proj-path resource))
+                    (prn 'disk only-in-disk-pb-msg)
+                    (prn 'save only-in-save-pb-msg)))
+
+                ;; We don't have a read-fn. Compare the strings.
+                (let [disk-string (slurp resource)]
+                  (is (= disk-string save-string)))))))))))
```

Comment:
This is another test that was testing the wrong thing by being too low-level. Rewrote it to use the `read-fn` (and by extension, perform data sanitation) instead of calling `protobuf/read-text` directly. Sanitation is desired, because it means we don't have to continuously update the reference files if we add fields, etc.